### PR TITLE
Add in-memory dataset data pipe

### PR DESCRIPTION
**Pull request by ys-teh**
# ALCHEMI Toolkit Pull Request

## Description

Add an in-memory dataset data pipe for datasets that fit in CPU memory, integrating reader-backed resident `Batch` materialization with the existing ALCHEMI `DataLoader` fused-prefetch path and DDP sampler setup. This also exposes an explicit neighbor-list method override through model and pipeline neighbor-hook creation.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

N/A

## Changes Made

- `InMemoryDataset`
  - Added a resident-batch data pipe that materializes reader-backed data into memory, with optional validation skipping, device transfer, fused batch prefetch, and context management.
  - Integrated `InMemoryDataset` with the existing ALCHEMI `DataLoader` batch contract and DDP default distributed sampler path.
  - Documented the new data pipe, including user guide examples and changelog notes.
  - Added targeted regression coverage for in-memory loading, fused prefetch, DataLoader behavior, and DDP integration.
- Neighbor list method selection: allowed callers to pass an explicit `neighbor_list_method` from model and pipeline neighbor-hook creation into `NeighborListHook`, with regression coverage for forwarding.

## Testing

Targeted tests run:

- `uv run --extra cu12 --extra mace pytest test/data/test_zarr_datapipe.py test/hooks/test_neighbor_list_hook.py test/models/test_base.py test/models/test_pipeline.py test/training/test_ddp_hook.py` (passed: 540 passed, 15 warnings)

- [ ] Unit tests pass locally (`make pytest`)
- [ ] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

A number of class methods such as `_resolve_target_device` and `get_fused_batches` are copied from `Dataset`. Recommend creating shared utility functions in the future.

### MACE training benchmark

These runs use the MACE training example with batch size 256, 8 CPUs, 1 GPU, and `skip_validation=True`. The averaged profiler columns use 990 measured training steps, excluding the first 10 steps. Rows 1-4 use the existing `Dataset` path with different prefetch factors. Rows 5-6 show the impact of `InMemoryDataset` on data loading and end-to-end runtime. Row 7 adds explicit neighbor-list method selection, reducing the before-forward hook time, which is dominated by neighbor-list construction in this workload.

| Configuration | Slurm elapsed | Training run (s) | Batch avg (ms) | Data loading avg (ms) | Before-forward hooks avg (ms) |
| --- | ---: | ---: | ---: | ---: | ---: |
| `Dataset`, prefetch 4 | 00:06:03 | 330.626 | 134.994 | 120.358 | 27.058 |
| `Dataset`, prefetch 128 | 00:06:27 | 344.255 | 115.078 | 82.093 | 28.791 |
| `Dataset`, prefetch 512 | 00:08:59 | 513.106 | 83.278 | 78.866 | 16.392 |
| `Dataset`, prefetch 4096 | 00:08:03 | 454.566 | 65.321 | 0.453 | 8.758 |
| `InMemoryDataset`, CPU-resident | 00:03:58 | 142.020 | 69.292 | 1.500 | 11.413 |
| `InMemoryDataset`, async H2D prefetch 2 | 00:03:54 | 144.222 | 63.167 | 6.034 | 8.234 |
| `InMemoryDataset`, async H2D prefetch 2, explicit neighbor-list method | 00:03:56 | 133.569 | 58.445 | 1.742 | 2.980 |

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemi/data/datapipes/in_memory_dataset.py [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+        pending = self._prepare_batches(batch_index_lists, stream=stream)
+        if pending.event is not None:
+            pending.event.synchronize()
+        return pending.batches
+
+    def prefetch_fused_batches(
+        self,
+        batch_index_lists: Sequence[Sequence[int]] | Sequence[Tensor],
+        stream: torch.cuda.Stream | None = None,
+    ) -> None:
```
[the remark this answers is not in the document]
Comment (ys-teh) [reply nvalchemi/data/datapipes/in_memory_dataset.py:471]: [Addressed](https://github.com/NVIDIA/nvalchemi-toolkit/pull/126/commits/0f047be39b4b2bc30699694b2faf59892c9d65ce) and added batch_transform support for in_memory_batch provided by user.

File: nvalchemi/data/datapipes/in_memory_dataset.py [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+            If ``batch_transforms`` is not a
+            :class:`~collections.abc.Sequence` (e.g. a single callable or a
+            generator was passed).
+        """
+        if (in_memory_batch is None) == (reader is None):
+            raise ValueError("Pass exactly one of in_memory_batch or reader.")
+        self.target_device = self._resolve_target_device(device)
+        if in_memory_batch is None:
+            if reader is None:
+                raise ValueError("reader must be provided when in_memory_batch is None")
```
[the remark this answers is not in the document]
Comment (ys-teh) [reply nvalchemi/data/datapipes/in_memory_dataset.py]: Reduced to one if-statement

File: nvalchemi/training/hooks/ddp.py [hunk 3]
```diff
@@ -414,7 +417,7 @@ def _prepare_nvalchemi_dataloader(
                 batch_size=dataloader.batch_size,
                 **kwargs,
             )
-        elif isinstance(dataset, ALCHEMIDataset):
+        elif isinstance(dataset, (ALCHEMIDataset, ALCHEMIInMemoryDataset)):
```
Comment (laserkelvin) [inline nvalchemi/training/hooks/ddp.py]: I think this is probably a one-off addition, but I'm wondering whether we should just have this checked as a protocol that will work for both

File: test/data/test_zarr_datapipe.py
Comment (laserkelvin) [inline test/data/test_zarr_datapipe.py:1]: Can you add test cases where we use a different sampler than the default, just to make sure that interface works?

Ideally the distributed sampler, but any custom sampler should be good enough, and have it work with the dataloader

**Review by laserkelvin — CHANGES_REQUESTED** [review]
Generally LGTM

- If the dataset is small (on the order of a few GB) and will fit in host memory, we should use `InMemory`)` but if it doesn't they should fall back to regular dataset - update the agent skill to make a suggestion like this
- I'm also interested to know if we can mix and match in-memory and regular datasets when using `MultiDataset` - if you can add a unit test case for that it would be great. I don't foresee either cases not working but good to test against

File: nvalchemi/training/hooks/ddp.py (hunk 3 above)
Comment (ys-teh) [reply nvalchemi/training/hooks/ddp.py]: Good idea. I added `BatchDatasetProtocol` and also additional class methods like `field_names`, `prefetch`, `prefetch_count`, `get_metadata` in `InMemoryDataset` for consistency and compatibility. `DataLoader` and `DDPHook` now use the protocol where they handle single batch-loadable datasets, while `MultiDataset` accepts protocol-compatible datasets as inputs.

File: test/data/test_zarr_datapipe.py
Comment (ys-teh) [reply test/data/test_zarr_datapipe.py:1]: Yes, added `test_in_memory_dataloader_distributed_sampler`, mimicking existing `test_dataloader_distributed_sampler`.

**Comment by ys-teh** [comment]
> Generally LGTM
> 
> * If the dataset is small (on the order of a few GB) and will fit in host memory, we should use `InMemory`)` but if it doesn't they should fall back to regular dataset - update the agent skill to make a suggestion like this
> * I'm also interested to know if we can mix and match in-memory and regular datasets when using `MultiDataset` - if you can add a unit test case for that it would be great. I don't foresee either cases not working but good to test against

Thanks. I updated the skills [here](https://github.com/NVIDIA/nvalchemi-toolkit/pull/126/commits/726ec62ba86d3dbfa4f91882375fc2351b839db9).

Yes, confirmed the mix-and-match compatibility [here](https://github.com/NVIDIA/nvalchemi-toolkit/pull/126/commits/bc3cd7ea139bea58c40b1fdcc3ab51ca4f6a4f1f) with `test_multidataset_accepts_dataset_and_in_memory_dataset`.

File: nvalchemi/data/datapipes/in_memory_dataset.py [hunk 4]
```diff
@@ -467,6 +494,21 @@ def prefetch_fused_batches(
...
+        there is no reader I/O to overlap for an individual sample.
+
+        Parameters
+        ----------
+        index : int
+            Graph index within ``in_memory_batch``.
+        stream : torch.cuda.Stream | None, default=None
+            Unused CUDA stream argument retained for compatibility.
+        """
+        del index, stream
```
Comment (laserkelvin) [inline nvalchemi/data/datapipes/in_memory_dataset.py]: Why do you need to `del` them?

File: nvalchemi/training/hooks/ddp.py [hunk 5]
```diff
@@ -417,7 +416,7 @@ def _prepare_nvalchemi_dataloader(
                 batch_size=dataloader.batch_size,
                 **kwargs,
             )
-        elif isinstance(dataset, (ALCHEMIDataset, ALCHEMIInMemoryDataset)):
+        elif isinstance(dataset, ALCHEMIBatchDatasetProtocol):
```
Comment (laserkelvin) [inline nvalchemi/training/hooks/ddp.py]: Probably don't need to prefix with `ALCHEMI` - the others were done just because of the namespace collision, which this won't happen for (it just looks verbose IMHO)

**Review by laserkelvin — APPROVED** [review]
Minor comments - we can merge once you update the branch

File: nvalchemi/data/datapipes/in_memory_dataset.py (hunk 4 above)
Comment (ys-teh) [reply nvalchemi/data/datapipes/in_memory_dataset.py]: Looks like there is no issue with unused variables. Updated to return None.

File: nvalchemi/training/hooks/ddp.py (hunk 5 above)
Comment (ys-teh) [reply nvalchemi/training/hooks/ddp.py]: Updated.

**Review by laserkelvin — APPROVED** [review]
Approved - thanks for making the changes!

**Comment by laserkelvin** [comment]
/ok to test [229bf3d](https://github.com/NVIDIA/nvalchemi-toolkit/pull/126/commits/229bf3dd5adffdead34e26f59e924e66fe37353b)
